sdk/extensions/Microsoft.Extensions.Azure/tests/AzureClientFactoryTests.cs

```diff
@@ -417,9 +450,158 @@ public void CanRegisterCustomClientWithOptionsAndCredentialProvider()
+        }
+
+        [Test]
+        public void DisposeHandlesNonDisposableClients()
+        {
+            var services = new ServiceCollection();
+
+            services.AddAzureClients(builder =>
+                builder.AddClient<NonDisposableClient, object>(_ => new NonDisposableClient()).WithName(nameof(NonDisposableClient)));
+
+            var provider = services.BuildServiceProvider();
+            var factory = provider.GetRequiredService<IAzureClientFactory<NonDisposableClient>>();
+            var client = factory.CreateClient(nameof(NonDisposableClient));
+
+            Assert.DoesNotThrow(provider.Dispose);
+        }
+
+        [Test]
+        public void DisposableWorksWithMixedClients()
+        {
+            var tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var services = new ServiceCollection();
+            var disposeCount = 0;
+
+            Action disposeCallback= () =>
+            {
+                if (++disposeCount >=2)
+                {
+                    tcs.TrySetResult(true);
+                }
+            };
+
+            services.AddAzureClients(builder =>
+                builder.AddClient<NonDisposableClient, object>(_ => new NonDisposableClient()).WithName(nameof(NonDisposableClient)));
+
+            services.AddAzureClients(builder =>
+                builder.AddClient<AsyncDisposableClient, object>(_ => new AsyncDisposableClient(disposeCallback)).WithName(nameof(AsyncDisposableClient)));
+
+            services.AddAzureClients(builder =>
+                builder.AddClient<DisposableClient, object>(_ => new DisposableClient(disposeCallback)).WithName(nameof(DisposableClient)));
+
+            var provider = services.BuildServiceProvider();
+
+            var factory = provider.GetRequiredService<IAzureClientFactory<NonDisposableClient>>();
+            var client = factory.CreateClient(nameof(NonDisposableClient));
+
+            var disposableFactory = provider.GetRequiredService<IAzureClientFactory<DisposableClient>>();
+            var disposableClient = disposableFactory.CreateClient(nameof(DisposableClient));
+
+            var asyncDisposableFactory = provider.GetRequiredService<IAzureClientFactory<AsyncDisposableClient>>();
+            var asyncDispsableClient = asyncDisposableFactory.CreateClient(nameof(AsyncDisposableClient));
+
+            Assert.DoesNotThrow(provider.Dispose);
+
+            using var cancellationSource = new CancellationTokenSource(TimeSpan.FromSeconds(5));
+
+            Assert.DoesNotThrowAsync(async () => await tcs.Task.AwaitWithCancellation(cancellationSource.Token));
+            Assert.AreEqual(disposeCount, 2);
+        }
+
         private IConfiguration GetConfiguration(params KeyValuePair<string, string>[] items)
         {
             return new ConfigurationBuilder().AddInMemoryCollection(items).Build();
         }
+
+        private class DisposableClient : IDisposable
```

Comment:
Since we have logic around types that are both IDisposable and IAsyncDisposable, should we have a test case that covers this?

Reply:
The version of the DI package that we reference has a service factory that only supports Dispose, while the newer versions support both.   I could artificially test by retrieving each factory that we registered and manually invoking DisposeAsync, but I didn't feel that it offered enough value to justify the effort.

Reply:
Bumped the dependency to allow access to DisposeAsync and mirrored tests